test/test_apply_layers.py

```diff
@@ -0,0 +1,273 @@
+    output = apply_layers(layers, input_data.clone())
+    super().compareResults(output, input_data, abs_err=0.0001, rel_err=0.001)
+
+  def test_linear_layers(self):
+    # Fix the random seed to avoid flakes.
+    with torch.random.fork_rng():
+      with torch_xla.xm.fork_rng():
+        torch.random.manual_seed(42)
+        torch_xla.xm.set_rng_state(42)
+        # We want to apply these layers sequentially
+        layers = [nn.Linear(64, 64).to(self.device) for _ in range(10)]
+        input_data = torch.randn(64).to(self.device)
+        torch_xla.sync(wait=True)
+
+    scan_layers = deepcopy(layers)
+    loop_layers = deepcopy(layers)
+    torch_xla.sync()
+
+    output = apply_layers(scan_layers, input_data.clone())
+    output.sum().backward()
+
+    # Test that the result is the same as for loop.
+    loop_output = input_data.clone()
+    for layer in loop_layers:
+      loop_output = layer(loop_output)
+    torch_xla.sync()
+
+    super().compareResults(loop_output, output, abs_err=0.0001, rel_err=0.001)
+    self.assert_different_tensor(loop_output, output)
+
+    loop_output.sum().backward()
+    torch_xla.sync()
+
+    # Test that the gradients are the same too.
+    for layer_scan, layer_loop in zip(scan_layers, loop_layers):
+      assert layer_scan.weight.grad is not None
+      assert layer_loop.weight.grad is not None
+      assert layer_scan.bias.grad is not None
+      assert layer_loop.bias.grad is not None
+      super().compareResults(
+          layer_scan.weight.grad,
+          layer_loop.weight.grad,
+          abs_err=0.0001,
+          rel_err=0.001)
+      super().compareResults(
+          layer_scan.bias.grad,
+          layer_loop.bias.grad,
+          abs_err=0.0001,
+          rel_err=0.001)
+      self.assert_different_tensor(layer_scan.weight.grad,
+                                   layer_loop.weight.grad)
+      self.assert_different_tensor(layer_scan.bias.grad, layer_loop.bias.grad)
+
+  def test_tuple_layers(self):
+    """Test applying layers that consume and return tuples. Construct a module
+    that transforms each element in the tuple.
+    """
+
+    class TupleModule(torch.nn.Module):
+
+      def __init__(self):
+        super(TupleModule, self).__init__()
+        self.linear = nn.Linear(64, 64)
+        self.w = nn.Parameter(torch.randn(64, 64, requires_grad=True))
+
+      def forward(self, x, y, z):
+        return self.linear(x).sin(), self.linear(
+            y).cos(), self.linear(z) @ self.w
+
+    layers = [TupleModule().to(self.device) for _ in range(10)]
+    torch_xla.sync()
+
+    scan_layers = deepcopy(layers)
+    loop_layers = deepcopy(layers)
+    torch_xla.sync()
+
+    # Also make input data some non-trivial graph instead of just device data.
+    input_data = (torch.randn(64).to(self.device) * 100,
+                  torch.randn(64).to(self.device) * 200,
+                  torch.randn(64).to(self.device) * 300)
+    a = torch.randn(64).to(self.device)
```

Comment:
hmm, we need to check if this `input is an IR` will work in the SPMD case, it is a bit complicated we can chat more offline..

Reply:
*nod* i think this is a good point